Make Sygus conjectures higher-order

ajreynol opened the pull request:
This removes a long standing hack for representing SyGuS conjectures.

A SyGuS conjecture of the form:
exists f. forall x. P( f, x )

was previously represented as:
(EXISTS ((proxy T)) (FORALL ((x ...)) [P( f, x )]))
where proxy.getAttribute(SygusFunAttribute()) = f, and T is a first-order type that encodes the grammar restrictions.
Notice that f is a *free* variable in the above formula.

Due to our new support for higher-order, we no longer need this hack, and can represent this as:
(EXISTS ((f ...)) (FORALL ((x ...)) [P( f, x )]))
where f.getAttribute(SygusFunGrammar()) = proxy, where proxy.getType() encodes the grammar restrictions.

This fixes #1205.

This fixes a bug in node.cpp related to computing hasBoundVars (introduced while we were merging #1206).  This function needs to look for bound variables in the operator, which now can be the case.  Currently, a SyGuS benchmark family is failing in master cvc4 because of this. I've added a regression from this family.

ajreynol commented:
This matches the performance from a few weeks ago (before the bug was introduced in PR #1206) : https://www.starexec.org/starexec/secure/details/job.jsp?id=23847
See CVC4-101317-2-sygusHo

timothy-king commented on src/parser/smt2/smt2.cpp:
```diff
@@ -1081,13 +1081,16 @@ const void Smt2::getSygusPrimedVars( std::vector<Expr>& vars, bool isPrimed ) {
+  //proxy variable carries the type, which may be an SyGuS datatype
+  //that corresponds to syntactic restrictions.
+  Expr sym = mkBoundVar("sfproxy", t);  
   std::vector< Expr > attr_value;
```
inline attr_value.
> Command* cattr = new SetUserAttributeCommand("sygus-synth-grammar", synth_fun, {sym});

timothy-king commented on src/smt/smt_engine.cpp:
```diff
@@ -4549,10 +4549,7 @@ Result SmtEngine::checkSynth(const Expr& e) throw(Exception) {
       quantifiers::SingleInvocationPartition sip;
       std::vector< Node > funcs;
```
const std::vector<Node> funcs{conj[0].begin(), conj[0].end()};

timothy-king commented on src/parser/smt2/smt2.cpp:
```diff
@@ -1081,13 +1081,16 @@ const void Smt2::getSygusPrimedVars( std::vector<Expr>& vars, bool isPrimed ) {
-  //FIXME #1205 : we should not create a proxy, instead quantify on synth_fun and set Type t as an attribute
-  Expr sym = mkBoundVar("sfproxy", t);
-  d_sygusFunSymbols.push_back(sym);
+  // when constructing the synthesis conjecture, we quantify on the 
```
Capitalize "When"

timothy-king commented on src/parser/smt2/smt2.cpp:
```diff
@@ -1081,13 +1081,16 @@ const void Smt2::getSygusPrimedVars( std::vector<Expr>& vars, bool isPrimed ) {
-  Expr sym = mkBoundVar("sfproxy", t);
-  d_sygusFunSymbols.push_back(sym);
+  // when constructing the synthesis conjecture, we quantify on the 
+  // (higher-order) bound variable synth_fun
```
Add a period to the end of the sentence.

timothy-king commented on src/parser/smt2/smt2.cpp:
```diff
@@ -1081,13 +1081,16 @@ const void Smt2::getSygusPrimedVars( std::vector<Expr>& vars, bool isPrimed ) {
+  // (higher-order) bound variable synth_fun
+  d_sygusFunSymbols.push_back(synth_fun);
   
+  //proxy variable carries the type, which may be an SyGuS datatype
```
Ditto

ajreynol commented:
I've addressed the review.  I do not want to use the new syntax for vectors.  The older syntax is clearer in my opinion.  But we can discuss this as a style guidelines if you want.
